<a href="https://colab.research.google.com/github/mauriani/trabalho-ia-01/blob/main/Trabalho_DE_IA_01_QUEST%C3%83O_A.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import requests

data_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/communities/communities.data'
columns_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/communities/communities.names'

response = requests.get(columns_url)
columns = []


for line in response.text.splitlines():
    # pega os nomes das colunas depois do '@attribute'
    if line.startswith('@attribute'):
        column_name = line.split()[1]
        columns.append(column_name)


ds = pd.read_csv(data_url, header=None, names=columns, na_values='?')

ds.head()


state  county  community        communityname  fold  population  \
0      8     NaN        NaN         Lakewoodcity     1        0.19   
1     53     NaN        NaN          Tukwilacity     1        0.00   
2     24     NaN        NaN         Aberdeentown     1        0.00   
3     34     5.0    81440.0  Willingborotownship     1        0.04   
4     42    95.0     6096.0    Bethlehemtownship     1        0.01   

   householdsize  racepctblack  racePctWhite  racePctAsian  ...  LandArea  \
0           0.33          0.02          0.90          0.12  ...      0.12   
1           0.16          0.12          0.74          0.45  ...      0.02   
2           0.42          0.49          0.56          0.17  ...      0.01   
3           0.77          1.00          0.08          0.12  ...      0.02   
4           0.55          0.02          0.95          0.09  ...      0.04   

   PopDens  PctUsePubTrans  PolicCars  PolicOperBudg  LemasPctPolicOnPatr  \
0     0.26            0.20       0.06           0.04                  0.9   
1     0.12            0.45        NaN            NaN                  NaN   
2     0.21            0.02        NaN            NaN                  NaN   
3     0.39            0.28        NaN            NaN                  NaN   
4     0.09            0.02        NaN            NaN                  NaN   

   LemasGangUnitDeploy  LemasPctOfficDrugUn  PolicBudgPerPop  \
0                  0.5                 0.32             0.14   
1                  NaN                 0.00              NaN   
2                  NaN                 0.00              NaN   
3                  NaN                 0.00              NaN   
4                  NaN                 0.00              NaN   

   ViolentCrimesPerPop  
0                 0.20  
1                 0.67  
2                 0.43  
3                 0.12  
4                 0.03  

[5 rows x 128 columns]

In [ ]:
def avaliaQualidadeDataframe(df):
  if df.isnull().values.any():
    print('Valores nulos encontrados.')
  else:
    print('Sem valores nulos.')
  for col in df.dtypes.to_list():
    if str(col) not in ["int64", "float64"]:
      print('Coluna com tipo incorreto para análise.')

In [ ]:
# Checa qualidade
avaliaQualidadeDataframe(ds)

Valores Nulos Encontrados.
Coluna com tipo incorreto para análise.


In [ ]:
# Removendo colunas com valores faltantes
ds_clean = ds.dropna(axis=0)

avaliaQualidadeDataframe(ds_clean)


Sem valores nulos.
Coluna com tipo incorreto para análise.


In [ ]:
# Removendo colunas NÃO númericas que não contribuiu para a regressão
# exemplo de coluna que está sendo removida é a de communityname

data = ds_clean.select_dtypes(include=['float64', 'int64'])

data.head()

state  county  community  fold  population  householdsize  racepctblack  \
16     36     1.0     1000.0     1        0.15           0.31          0.40   
23     19   193.0    93926.0     1        0.11           0.43          0.04   
33     51   680.0    47672.0     1        0.09           0.43          0.51   
68     34    23.0    58200.0     1        0.05           0.59          0.23   
74      9     9.0    46520.0     1        0.08           0.39          0.08   

    racePctWhite  racePctAsian  racePctHisp  ...  LandArea  PopDens  \
16          0.63          0.14         0.06  ...      0.06     0.39   
23          0.89          0.09         0.06  ...      0.16     0.12   
33          0.58          0.04         0.01  ...      0.14     0.11   
68          0.39          0.09         1.00  ...      0.01     0.73   
74          0.85          0.04         0.25  ...      0.07     0.21   

    PctUsePubTrans  PolicCars  PolicOperBudg  LemasPctPolicOnPatr  \
16            0.84       0.06           0.06                 0.91   
23            0.07       0.04           0.01                 0.81   
33            0.19       0.05           0.01                 0.75   
68            0.28       0.00           0.02                 0.64   
74            0.04       0.02           0.01                 0.70   

    LemasGangUnitDeploy  LemasPctOfficDrugUn  PolicBudgPerPop  \
16                  0.5                 0.88             0.26   
23                  1.0                 0.56             0.09   
33                  0.0                 0.60             0.10   
68                  0.0                 1.00             0.23   
74                  1.0                 0.44             0.11   

    ViolentCrimesPerPop  
16                 0.49  
23                 0.63  
33                 0.31  
68                 0.50  
74                 0.14  

[5 rows x 127 columns]

In [ ]:
# Nós queremos prever a váriavel ViolentCrimesPerPop para o nosso modelo de regressão linear

X = data.drop(columns=['ViolentCrimesPerPop'])
y = data['ViolentCrimesPerPop']



In [ ]:
# 1.B

# Divisao em treino e teste

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error


# criando o modelo e o treino
model = LinearRegression()
model.fit(X_train, y_train)


# Faz previções

y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

# Agora vamos fazer as métricas RMSE (Root Mean Square Error) e MAE (Mean Absolute Error) que sao
# usadas ara avaliar a precisao de modelos de previsao

rmse_train = mean_squared_error(y_train, y_train_pred, squared=False)
mae_train = mean_absolute_error(y_train, y_train_pred)

rmse_test = mean_squared_error(y_test, y_test_pred, squared=False)
mae_test = mean_absolute_error(y_test, y_test_pred)

print(f'RMSE (Root Mean Square Error): {rmse_train}, MAE: {mae_train}')
print(f'MAE (Mean Absolute Error: {rmse_test}, MAE: {mae_test}')


Treino de RMSE (Root Mean Square Error): 3.822997631204621e-12, MAE: 3.0430136769572306e-12
Teste de MAE (Mean Absolute Error: 0.31166911159576466, MAE: 0.23313523845982995


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


In [ ]:
# 1.C

# Aplicando PCA

from sklearn.decomposition import PCA

pca = PCA(n_components=5)
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)

print(f'Variancia - {pca.explained_variance_ratio_}')

Variancia - [9.99934666e-01 6.51135848e-05 2.00773831e-07 1.30273797e-08
 2.21503368e-09]


In [ ]:
model_pca = LinearRegression()
model_pca.fit(X_train_pca, y_train)

y_train_pred_pca = model_pca.predict(X_train_pca)
y_test_pred_pca = model_pca.predict(X_test_pca)

# Avaliar o modelo com PCA
rmse_train_pca = mean_squared_error(y_train, y_train_pred_pca, squared=False)
mae_train_pca = mean_absolute_error(y_train, y_train_pred_pca)

rmse_test_pca = mean_squared_error(y_test, y_test_pred_pca, squared=False)
mae_test_pca = mean_absolute_error(y_test, y_test_pred_pca)

print(f'Treino de RMSE COM PCA: {rmse_train}, MAE: {mae_train}')
print(f'Teste de MAE COM PCA: {rmse_test}, MAE: {mae_test}')


Treino de RMSE COM PCA: 3.822997631204621e-12, MAE: 3.0430136769572306e-12
Teste de MAE COM PCA: 0.31166911159576466, MAE: 0.23313523845982995


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


# CONCLUSAO

# RESULTADO SEM PCA
Treino de RMSE : 3.822997631204621e-12, MAE: 3.0430136769572306e-12
Teste de MAE : 0.31166911159576466, MAE: 0.23313523845982995

# RESULTADO COM PCA
Treino de RMSE COM PCA: 3.822997631204621e-12, MAE: 3.0430136769572306e-12
Teste de MAE COM PCA: 0.31166911159576466, MAE: 0.23313523845982995


# Análise
### TREINO RMSE E MAE
Podemos observar que os dados RMSE e o MAE de ambos tiveram o mesmo valor, isso é um sinal
de overfitting. O overfitting no treino torna-se preocupante caso ocorra muitas váriaveis ou se os dados possuem redundãncias.

### ANALISE DE VALORES COM E SEM PCA

OS valores apresentados mostram que os modelos possuiem desempenho semelhantes/iguais
e que o PCA não imapctou a perfomance do conjunto de dados.
